kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java

```diff
@@ -164,6 +178,168 @@ public void testShouldHandleRecoveryTask() throws Exception {
+            "true",
+            "500",
+            "500",
+            components.logMessageWriter,
+            components.meterRegistry,
+            components.consumerOverrideProps);
+    final Instant startTime =
+        LocalDateTime.of(2020, 10, 1, 10, 10, 0).atZone(ZoneOffset.UTC).toInstant();
+    final long msgsToProduce = 100;
+    TestKafkaServer.produceMessagesToKafka(
+        components.testKafkaServer.getBroker(),
+        startTime,
+        topicPartition.topic(),
+        topicPartition.partition(),
+        (int) msgsToProduce);
+    await().until(() -> localTestConsumer.getEndOffSetForPartition() == msgsToProduce);
+    setRetentionTime(components.adminClient, topicPartition.topic(), 250);
+    with()
+        .atMost(1, TimeUnit.MINUTES)
+        .await()
+        .until(() -> getStartOffset(components.adminClient, topicPartition) > 0);
+
+    // produce some more messages that won't be expired
+    setRetentionTime(components.adminClient, topicPartition.topic(), 25000);
+    TestKafkaServer.produceMessagesToKafka(
+        components.testKafkaServer.getBroker(),
+        startTime,
+        topicPartition.topic(),
+        topicPartition.partition(),
+        (int) msgsToProduce);
+    await()
+        .until(() -> localTestConsumer.getEndOffSetForPartition() == msgsToProduce + msgsToProduce);
+
+    SnapshotMetadataStore snapshotMetadataStore = new SnapshotMetadataStore(metadataStore, false);
+    assertThat(snapshotMetadataStore.listSync().size()).isZero();
+
+    // Start recovery service
+    recoveryService =
+        new RecoveryService(kaldbCfg, metadataStore, components.meterRegistry, blobFs);
+    recoveryService.startAsync();
+    recoveryService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    // Start recovery with an offset range that is partially unavailable
+    long startOffset = 50;
+    long endOffset = 150;
+    RecoveryTaskMetadata recoveryTask =
+        new RecoveryTaskMetadata(
+            topicPartition.topic(),
+            Integer.toString(topicPartition.partition()),
+            startOffset,
+            endOffset,
+            Instant.now().toEpochMilli());
+    assertThat(recoveryService.handleRecoveryTask(recoveryTask)).isTrue();
+    assertThat(getCount(RECORDS_NO_LONGER_AVAILABLE, components.meterRegistry)).isEqualTo(50);
+    assertThat(getCount(MESSAGES_RECEIVED_COUNTER, components.meterRegistry)).isEqualTo(51);
+    List<SnapshotMetadata> snapshots = snapshotMetadataStore.listSync();
+    assertThat(snapshots.size()).isEqualTo(1);
+    assertThat(blobFs.listFiles(BlobFsUtils.createURI(TEST_S3_BUCKET, "/", ""), true)).isNotEmpty();
+    assertThat(blobFs.exists(URI.create(snapshots.get(0).snapshotPath))).isTrue();
+    assertThat(blobFs.listFiles(URI.create(snapshots.get(0).snapshotPath), false).length)
+        .isGreaterThan(1);
+    assertThat(getCount(MESSAGES_FAILED_COUNTER, meterRegistry)).isEqualTo(0);
+    assertThat(getCount(ROLLOVERS_INITIATED, meterRegistry)).isEqualTo(0);
```

Comment:
Something is off here. It seems we have snapshots without roll overs. Is the chunk with 51 messages rolled over?

Reply:
No, the chunk manager is configured to roll at either 10GB or 10k messages.